### ISSUE TITLE
Configuration error for FIPS in Alpine Image

### ISSUE THREAD

**Issue by sharathkramadas** [openssl/openssl#20595]
I have been trying to enable FIPS for OpenSSL for one of my alpine docker image

```
RUN wget --quiet https://www.openssl.org/source/openssl-3.1.0.tar.gz

ENV OPENSSL_FIPS=1

RUN apk add --no-cache zlib \
  && apk add --no-cache --virtual .build-deps \
      wget \
      gcc \
      gzip \
      tar \
      libc-dev \
      ca-certificates \
      perl \
      make \
      coreutils \
      linux-headers \
      zlib-dev \
  && tar -xzf openssl-3.1.0.tar.gz \
  && cd openssl-3.1.0 \
  && ./config \
  && make \
  && make install \
  && cd .. \
  && cd openssl-3.1.0 \
  && ./Configure enable-fips \
  && make \
  && make install
```

But after the installation i don't see FIPS being enabled

```
/tmp/build # openssl version
OpenSSL 3.1.0 14 Mar 2023 (Library: OpenSSL 3.0.8 7 Feb 2023)

/tmp/build # echo "test" | openssl md5
MD5(stdin)= d8e8fca2dc0f896fd7cb4cb0031ba249
```

Am i missing something?

**Comment by mattcaswell** [comment]
> OpenSSL 3.1.0 14 Mar 2023 (Library: OpenSSL 3.0.8 7 Feb 2023)

By the way this indicates a mis-configuration somewhere. The command line openssl application and the library version are different. Probably you are picking up the system OpenSSL library version.

**Comment by sharathkramadas** [comment]
Thanks @mattcaswell, That helped!

**Comment by paulidale** [comment]
You are also required to build OpenSSL exactly as per the security policy if you want a FIPS results.  Currently 3.1 is not FIPS validated.  The only validated FIPS provider is version 3.0.0 -- you should build & install this using the `enable-fips` configuration time option.  This can then be used with a non-FIPS OpenBuild (3.1 in you case).

**Comment by sharathkramadas** [comment]
@mattcaswell in the older OpenSSL version after we enable FIPS, the `openssl version` command would return an output very similar to this `OpenSSL 1.0.2k-fips`, Would the latest OpenSSL versions also give the same output, Because i'm not seeing it, Anything i had missed?

**Comment by t8m** [comment]
There are no different openssl builds with FIPS or without FIPS support. The only difference if you do not use enable-fips during the configure step is that the FIPS provider shared module won't be built. The libcrypto and libssl libraries will be the same though. For that reason there is no `-fips` in the version string.

**Comment by sharathkramadas** [comment]
Perfect! Thanks @t8m !

**Comment by t8m** [comment]
@fgarcialainez MD5 is not FIPS approved algorithm. You should try SHA256.

If you change your configuration to load the default provider does `openssl list -digest-commands` work as expected?

**Comment by fgarcialainez** [comment]
Hi @t8m,

First of all, thanks for your quick answer.

Yes yes, I know MD5 is not FIPS approved, it is why I am trying specifically this. In other tests I have done with 1.0.2 FIPS approved, I get an error in MD5 stating that this digest is not available in FIPS mode. I have tried also with SHA256 and others, but same error in all them.

If I disable FIPS provider digest commands works as expected. Also, if I set 
```
[algorithm_sect]
default_properties = fips=no
```
it works as expected too.

**Comment by t8m** [comment]
The error message is different from the FIPS module 2.0 version.

Can you please try:
```
./util/wrap.pl -fips openssl sha256 ./test.txt
```

I assume that the openssl is looking for the fips module at some different place than expected.

You could also use the `strace` command to see if there aren't any unexpected errors or files opened with incorrect path.

**Comment by fgarcialainez** [comment]
@t8m, ah ok thanks for letting me know about the message.

On the other hand, I understand that listing the available digest commands should return all available in FIPS mode (e.g. MD5 shouldn´t be in the list, but SHA256 should be), right?

I will try to run the proposed commands and will let you know the result.

Thanks so much for your quick help!

**Comment by fgarcialainez** [comment]
@t8m, 

After further verifications, I have to say that it also works really well with Python 3.10.13 without any further adaptation. On enabling FIPS mode hashlib md5() function raising an unsupported error, while it creates the hash properly when FIPS mode is disabled.

On the other hand, I have an issue on trying the same stuff with old versions of Python (e.g. 3.5.9). First expected issue I am finding is that Python don't take automatically the same OpenSSL version than I have build from code, it is still using default OpenSSL 1.1.1 available in the python:3.5-alpine image. I have tried to point libssl.so.3 to libssl.so.1.1 and libcrypto.so.3 to libcrypto.so.1.1, without success (also expecting this problem).

```
/ # python --version
Python 3.5.9
/ # python3 -c "import ssl; print(ssl.OPENSSL_VERSION)"
OpenSSL 1.1.1d  10 Sep 2019
/ # openssl version
OpenSSL 3.0.8 7 Feb 2023 (Library: OpenSSL 3.0.8 7 Feb 2023)
```
My question here is what is the recommended approach for having Python 3.5.X using an OpenSSL version that is FIPS compliant. Is there any potential way of making Python 3.5.X work with OpenSSL 3.0.X?

Other option that I am thinking about is that we could try to replace 1.1.1 default version used by Python 3.5.X by 1.0.2 and build the FIPS module 2.0 (or something like that). But not sure if all the Python packages and dependencies will work properly with this version.

Any suggestions or guidance would be welcome.

Thanks in advance.

**Comment by t8m** [comment]
Unfortunately I do not know answers to these python questions.

Just please note that the FIPS module 2.0 is out of support and the validation status for it is historic.

**Comment by fgarcialainez** [comment]
@t8m 

I know FIPS module 2.0 is out of support and OpenSSL 1.0.2 is really out-of-date. 

Also, I can see that older Python versions (in theory) don't have full support for OpenSSL 3.X. I can see specific threads to include support for this in newer Python versions (e.g. https://discuss.python.org/t/openssl-3-0-0-support-3-8-to-3-10/8205).

At this point, what options are available? I understand that there is no way to include FIPS Module in OpenSSL 1.1.1, right? 

Upgrading to OpenSSL 3.X would require to upgrade Python version, and downgrading to OpenSSL 1.0.2 seems that could break many things taking into account that this Alpine version comes with OpenSSL 1.1.1 out-of-the-box.

**Comment by t8m** [comment]
Unfortunately I do not really see any good options here.

**Comment by fgarcialainez** [comment]
@t8m, ok thanks for your answer. I will keep you posted if I achieve to get a working configuration. Regards.

**Comment by fgarcialainez** [comment]
@t8m, just as an update. Few weeks ago I achieved to build Python 3.5.9 with OpenSSL 1.0.2 with FIPS module 2.0, but without uninstalling / removing default OpenSSL 1.1.1 (we just keep the two OpenSSL versions). At first sight, this approach seems to work properly.

On the other hand, no way to link properly OpenSSL 3.X with Python 3.5. I tried to follow the same approach that with 1.0.2 version, but Python keeps pointing to default OpenSSL version.

This is enough for the scope of the POC I was driving. One more time, thanks for your quick help and answers. 

Kind Regards.
